CVLByExample/ERC20Assembly/ERC20UtilsEquivalence.spec

```diff
@@ -0,0 +1,297 @@
+    function mock.balance() external returns (uint256) envfree;
+
+    /// DISPATCHERS
+    function _.permit(address,address,uint256,uint256,uint8,bytes32,bytes32) external => DISPATCHER(true);
+    function _.permit(address,address,uint256,uint256,bool,uint8,bytes32,bytes32) external => DISPATCHER(true);
+    function _.transfer(address,uint256) external => DISPATCHER(true);
+    function _.transferFrom(address,address,uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    
+    function _._ external => DISPATCH [
+        mock.approve(address,uint256),
+    ] default NONDET;
+}
+
+definition permit224_sig() returns uint32 = 0xd505accf;
+    //sig:mock.permit(address,address,uint256,uint256,uint8,bytes32,bytes32).selector;
+
+definition permit256_sig() returns uint32 = 0x8fcbaf0c;
+    //sig:mock.permit(address,address,uint256,uint256,bool,uint8,bytes32,bytes32).selector;
+
+definition transfer_sig() returns uint32 = 0xa9059cbb;
+    //sig:mock.transfer(address,uint256).selector;
+
+definition transferFrom_sig() returns uint32 = 0x23b872dd;
+    //sig:mock.transferFrom(uint256,uint256,address,bytes).selector;
+
+definition balanceOf_sig() returns uint32 = 0x70a08231;
+    //sig:mock.balanceOf(address).selector;
+
+definition approve_sig() returns uint32 = 0x095ea7b3;
+    //sig:mock.approve(address,uint256).selector;
+
+/// ghost boolean for "function was called".
+persistent ghost bool called_permit224;
+persistent ghost bool called_permit256;
+persistent ghost bool called_transfer;
+persistent ghost bool called_transferFrom;
+persistent ghost bool called_balanceOf;
+persistent ghost bool called_approve;
+
+/// ghost address for function callee.
+persistent ghost address callee_token_permit224;
+persistent ghost address callee_token_permit256;
+persistent ghost address callee_token_transfer;
+persistent ghost address callee_token_transferFrom;
+persistent ghost address callee_token_balanceOf;
+persistent ghost address callee_token_approve;
+
+/// ghost boolean for "permit was successful"
+persistent ghost bool permit_succees;
+
+function nullify_ghosts() {
+    called_permit224 = false;
+    called_permit256 = false;
+    called_transfer = false;
+    called_transferFrom = false;
+    called_balanceOf = false;
+    called_approve = false;
+
+    callee_token_permit224 = 0;
+    callee_token_permit256 = 0;
+    callee_token_transfer = 0;
+    callee_token_transferFrom = 0;
+    callee_token_balanceOf = 0;
+    callee_token_approve = 0;
+
+    permit_succees = false;
+}
+
+hook STATICCALL(uint g, address addr, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc {
+    if (selector == balanceOf_sig()) {
+        called_balanceOf = true;
+        callee_token_balanceOf = addr;
+    }
+}
+
+hook CALL(uint g, address addr, uint value, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc {
+    if (selector == transfer_sig()) {
```

Comment:
this is nice and I just learned a few days ago about 'selector' in call hooks

Reply:
I see why: it's not obvious from the syntax that this keyword exists.